File: ballast-client/npm/src/signalr/signalr-client-event-subscriber.ts
import { HubConnection } from "@aspnet/signalr";
import { IDisposable, IEvent, IEventBus, TYPES as BallastCore, IApplicationEvent } from "ballast-core";
import { inject, injectable } from "inversify";
import { TYPES as BallastClient } from "./../dependency-injection/types";
import { ISignalRClientOptions } from "./signalr-client-options";
import { SignalRClientServiceBase } from "./signalr-client-service-base";

export interface ISignalRClientEventSubscriber extends IDisposable { }

@injectable()
export class SignalRClientEventSubscriber extends SignalRClientServiceBase implements ISignalRClientEventSubscriber {

    public static readonly hubName: string = "eventhub";
    public get hubName() {
        return SignalRClientEventSubscriber.hubName;
    }

    private readonly _eventBus: IEventBus;

    public constructor(
        @inject(BallastClient.SignalR.ISignalRClientOptions) serviceOptions: ISignalRClientOptions,
        @inject(BallastCore.Messaging.IEventBus) eventBus: IEventBus
    ) {
        super(serviceOptions);
        this._eventBus = eventBus;
    }

    protected afterSubscribe(hubConnection: HubConnection) {
        hubConnection.on("IApplicationEvent", this.onApplicationEvent);
    }

    protected beforeUnsubscribe(hubConnection: HubConnection) {
        hubConnection.off("IApplicationEvent");
    }
    
    private onApplicationEvent(evt: IApplicationEvent) {
        this._eventBus.publishAsync(evt); // Fire and forget
    }

}

File: ballast-client/dotnet/src/signalr/SignalRClientEventSubscriber.cs
using Ballast.Client.SignalR.Services;
using Ballast.Core.Application.Events;
using Ballast.Core.Messaging;
using Microsoft.AspNetCore.SignalR.Client;

namespace Ballast.Client.SignalR
{
    public interface ISignalRClientEventSubscriber { }
    public class SignalRClientEventSubscriber : SignalRClientServiceBase, ISignalRClientEventSubscriber
    {

        private readonly IEventBus _eventBus;

        public SignalRClientEventSubscriber(ISignalRClientOptions options, IEventBus eventBus) : base(options)
        {
            _eventBus = eventBus;
        }

        protected override string HubName => "eventhub";

        protected override void AfterSubscribe(HubConnection hubConnection)
        {
            hubConnection.On<IApplicationEvent>("IApplicationEvent", OnApplicationEvent);
        }

        protected override void BeforeUnsubscribe(HubConnection hubConnection)
        {
            hubConnection.Remove("IApplicationEvent");
        }

        private void OnApplicationEvent(IApplicationEvent evt)
        {
            _ = _eventBus.PublishAsync(evt); // Fire and forget
        }

    }
}